core/src/main/java/com/sequenceiq/cloudbreak/service/upgrade/ClusterUpgradeImageFilter.java

```diff
@@ -52,71 +46,86 @@ Images filter(List<Image> availableImages, Versions supportedVersions, Image cur
         return versionBasedImageFilter.getCdhImagesForCbVersion(supportedVersions, availableImages);
     }
 
-    private List<Image> filterImages(List<Image> availableImages, Image currentImage, String cloudPlatform) {
-        // I think the whole logic is broken, since we don't allow an upgrade if we shipped a new parcel extension...
-        return availableImages.stream()
+    private ImageFilterResult filterImages(List<Image> availableImages, Image currentImage, String cloudPlatform) {
+        List<Image> images = availableImages.stream()
+                .filter(filterCurrentImage(currentImage))
                 .filter(ignoredCmVersion())
-                .filter(validateCmVersion(currentImage).or(validateStackVersion(currentImage)))
+                .filter(validateCmAndStackVersion(currentImage))
                 .filter(validateCloudPlatform(cloudPlatform))
                 .filter(validateOsVersion(currentImage))
-                // We don't necessary need CFM and CSP since only SDX upgrade is supported
-                //.filter(validateCfmVersion(currentImage))
-                //.filter(validateCspVersion(currentImage))
                 .filter(validateSaltVersion(currentImage))
-                .filter(filterCurrentImage(currentImage))
                 .collect(Collectors.toList());
-    }
 
-    private Predicate<Image> validateOsVersion(Image currentImage) {
-        return image -> isOsVersionsMatch(currentImage, image);
+        return new ImageFilterResult(new Images(null, null, null, images, null), getReason(images));
     }
 
-    private boolean isOsVersionsMatch(Image currentImage, Image newImage) {
-        return newImage.getOs().equalsIgnoreCase(currentImage.getOs())
-                && newImage.getOsType().equalsIgnoreCase(currentImage.getOsType());
+    private Predicate<Image> filterCurrentImage(Image currentImage) {
+        return image -> {
+            boolean result = !image.getUuid().equals(currentImage.getUuid());
+            setReason(result, "No other image is available.");
```

Comment:
This message can be a little misleading, maybe there is other image but the input image id was incorrect?

Reply:
You're right. What about this? "Only your current image is available with the same package versions."

Reply:
Ok, pls. put in the imageId in the response.